File: plugins/openrouter/skills/model-routing-patterns/templates/routing-config.ts
/**
 * OpenRouter Model Routing Configuration (TypeScript)
 *
 * This module provides type-safe routing configuration for OpenRouter
 * with support for multiple routing strategies, fallback chains, and monitoring.
 */

export interface RoutingConfig {
  strategy: 'cost-optimized' | 'speed-optimized' | 'quality-optimized' | 'balanced' | 'custom';
  description: string;
  version: string;
  primary?: string;
  fallback?: string[];
  timeout: number;
  retry: RetryConfig;
  on_error: 'fallback' | 'retry' | 'fail';
  routing_rules?: Record<string, RoutingRule>;
  cost_tracking?: CostTrackingConfig;
  monitoring?: MonitoringConfig;
}

export interface RoutingRule {
  description: string;
  models: string[];
  fallback?: string[];
  classifier?: ClassifierConfig;
  max_tokens?: number;
  temperature?: number;
  streaming?: boolean;
  max_latency_ms?: number;
  timeout?: number;
}

export interface ClassifierConfig {
  conditions: string[];
}

export interface RetryConfig {
  max_attempts: number;
  delay_ms: number;
  exponential_backoff: boolean;
}

export interface CostTrackingConfig {
  enabled: boolean;
  log_requests?: boolean;
  alert_threshold_usd?: number;
  daily_budget_usd?: number;
}

export interface MonitoringConfig {
  enabled: boolean;
  metrics?: string[];
  alert_thresholds?: Record<string, number>;
}

/**
 * Model Router
 * Handles model selection based on routing configuration
 */
export class ModelRouter {
  private config: RoutingConfig;
  private requestCount: Map<string, number> = new Map();
  private totalCost: number = 0;

  constructor(config: RoutingConfig) {
    this.config = config;
  }

  /**
   * Select best model based on request context
   */
  selectModel(context: RequestContext): ModelSelection {
    // Apply routing rules
    if (this.config.routing_rules) {
      for (const [ruleName, rule] of Object.entries(this.config.routing_rules)) {
        if (this.matchesRule(context, rule)) {
          return {
            model: rule.models[0],
            fallback: rule.fallback || rule.models.slice(1),
            rule: ruleName,
            config: rule
          };
        }
      }
    }

    // Default to primary/fallback
    return {
      model: this.config.primary || 'anthropic/claude-4.5-sonnet',
      fallback: this.config.fallback || [],
      rule: 'default',
      config: {}
    };
  }

  /**
   * Check if request matches routing rule
   */
  private matchesRule(context: RequestContext, rule: RoutingRule): boolean {
    if (!rule.classifier) return false;

    for (const condition of rule.classifier.conditions) {
      // Simple condition evaluation (extend as needed)
      if (condition.includes('token_count <') && context.tokenCount) {
        const threshold = parseInt(condition.match(/\d+/)?.[0] || '0');
        if (context.tokenCount >= threshold) return false;
      }
      if (condition.includes('token_count >=') && context.tokenCount) {
        const threshold = parseInt(condition.match(/\d+/)?.[0] || '0');
        if (context.tokenCount < threshold) return false;
      }
    }

    return true;
  }

  /**
   * Execute request with fallback chain
   */
  async executeWithFallback(
    context: RequestContext,
    apiClient: OpenRouterClient
  ): Promise<ModelResponse> {
    const selection = this.selectModel(context);
    const models = [selection.model, ...(selection.fallback || [])];

    let lastError: Error | null = null;

    for (let i = 0; i < models.length; i++) {
      const model = models[i];

      try {
        console.log(`Attempting model ${i + 1}/${models.length}: ${model}`);

        const response = await this.executeWithRetry(
          model,
          context,
          apiClient,
          selection.config
        );

        // Track metrics
        this.trackRequest(model, response.cost);

        return response;

      } catch (error) {
        lastError = error as Error;
        console.error(`Model ${model} failed:`, error);

        if (i === models.length - 1) {
          // Last model in chain, throw error
          throw new Error(`All models failed. Last error: ${lastError.message}`);
        }

        // Continue to next model in fallback chain
        continue;
      }
    }

    throw new Error('Fallback chain exhausted');
  }

  /**
   * Execute request with retry logic
   */
  private async executeWithRetry(
    model: string,
    context: RequestContext,
    apiClient: OpenRouterClient,
    ruleConfig: Partial<RoutingRule>
  ): Promise<ModelResponse> {
    const maxAttempts = this.config.retry.max_attempts;
    let delay = this.config.retry.delay_ms;

    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        const response = await apiClient.chat({
          model,
          messages: context.messages,
          max_tokens: ruleConfig.max_tokens || 2000,
          temperature: ruleConfig.temperature || 0.7,
          stream: ruleConfig.streaming || false
        });

        return response;

      } catch (error) {
        if (attempt === maxAttempts) throw error;

        console.log(`Retry attempt ${attempt}/${maxAttempts} after ${delay}ms`);
        await this.sleep(delay);

        if (this.config.retry.exponential_backoff) {
          delay *= 2;
        }
      }
    }

    throw new Error('Max retries exceeded');
  }

  /**
   * Track request metrics
   */
  private trackRequest(model: string, cost: number) {
    const count = this.requestCount.get(model) || 0;
    this.requestCount.set(model, count + 1);
    this.totalCost += cost;

    // Check budget alerts
    if (this.config.cost_tracking?.alert_threshold_usd) {
      if (this.totalCost >= this.config.cost_tracking.alert_threshold_usd) {
        console.warn(`⚠️ Cost alert: $${this.totalCost.toFixed(2)} >= $${this.config.cost_tracking.alert_threshold_usd}`);
      }
    }
  }

  /**
   * Get routing statistics
   */
  getStats(): RoutingStats {
    return {
      totalRequests: Array.from(this.requestCount.values()).reduce((a, b) => a + b, 0),
      totalCost: this.totalCost,
      modelDistribution: Object.fromEntries(this.requestCount),
      avgCostPerRequest: this.totalCost / Array.from(this.requestCount.values()).reduce((a, b) => a + b, 0)
    };
  }

  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

/**
 * Types
 */
export interface RequestContext {
  messages: any[];
  tokenCount?: number;
  taskType?: string;
  complexityScore?: number;
  streamingRequired?: boolean;
  latencyRequirement?: number;
}

export interface ModelSelection {
  model: string;
  fallback: string[];
  rule: string;
  config: Partial<RoutingRule>;
}

export interface ModelResponse {
  content: string;
  model: string;
  cost: number;
  latency: number;
}

export interface RoutingStats {
  totalRequests: number;
  totalCost: number;
  modelDistribution: Record<string, number>;
  avgCostPerRequest: number;
}

export interface OpenRouterClient {
  chat(params: {
    model: string;
    messages: any[];
    max_tokens: number;
    temperature: number;
    stream: boolean;
  }): Promise<ModelResponse>;
}

/**
 * Example Usage
 */
export function exampleUsage() {
  // Load configuration
  const config: RoutingConfig = require('./balanced-routing.json');

  // Create router
  const router = new ModelRouter(config);

  // Mock API client
  const apiClient: OpenRouterClient = {
    async chat(params) {
      return {
        content: 'Response from ' + params.model,
        model: params.model,
        cost: 0.001,
        latency: 500
      };
    }
  };

  // Execute request
  const context: RequestContext = {
    messages: [{ role: 'user', content: 'Hello!' }],
    tokenCount: 100,
    taskType: 'classification'
  };

  router.executeWithFallback(context, apiClient)
    .then(response => {
      console.log('Response:', response);
      console.log('Stats:', router.getStats());
    })
    .catch(error => {
      console.error('Error:', error);
    });
}

File: plugins/openrouter/skills/model-routing-patterns/templates/routing-config.py
"""
OpenRouter Model Routing Configuration (Python)

This module provides model routing configuration for OpenRouter
with support for multiple routing strategies, fallback chains, and monitoring.
"""

import json
import time
from typing import Dict, List, Optional, Any, Tuple
from dataclasses import dataclass, field
from enum import Enum


class RoutingStrategy(Enum):
    """Available routing strategies"""
    COST_OPTIMIZED = "cost-optimized"
    SPEED_OPTIMIZED = "speed-optimized"
    QUALITY_OPTIMIZED = "quality-optimized"
    BALANCED = "balanced"
    CUSTOM = "custom"


@dataclass
class RetryConfig:
    """Retry configuration"""
    max_attempts: int = 3
    delay_ms: int = 1000
    exponential_backoff: bool = True


@dataclass
class RoutingRule:
    """Single routing rule"""
    description: str
    models: List[str]
    fallback: Optional[List[str]] = None
    classifier: Optional[Dict[str, Any]] = None
    max_tokens: int = 2000
    temperature: float = 0.7
    streaming: bool = False
    max_latency_ms: Optional[int] = None
    timeout: Optional[int] = None


@dataclass
class CostTrackingConfig:
    """Cost tracking configuration"""
    enabled: bool = True
    log_requests: bool = True
    alert_threshold_usd: float = 10.0
    daily_budget_usd: Optional[float] = None


@dataclass
class MonitoringConfig:
    """Monitoring configuration"""
    enabled: bool = True
    metrics: List[str] = field(default_factory=lambda: [
        "request_count", "success_rate", "avg_latency", "total_cost"
    ])
    alert_thresholds: Dict[str, float] = field(default_factory=dict)


@dataclass
class RoutingConfig:
    """Main routing configuration"""
    strategy: RoutingStrategy
    description: str
    version: str = "1.0.0"
    primary: Optional[str] = None
    fallback: Optional[List[str]] = None
    timeout: int = 5000
    retry: RetryConfig = field(default_factory=RetryConfig)
    on_error: str = "fallback"
    routing_rules: Optional[Dict[str, RoutingRule]] = None
    cost_tracking: Optional[CostTrackingConfig] = None
    monitoring: Optional[MonitoringConfig] = None


@dataclass
class RequestContext:
    """Request context for routing decisions"""
    messages: List[Dict[str, str]]
    token_count: Optional[int] = None
    task_type: Optional[str] = None
    complexity_score: Optional[float] = None
    streaming_required: bool = False
    latency_requirement: Optional[int] = None


@dataclass
class ModelSelection:
    """Selected model and configuration"""
    model: str
    fallback: List[str]
    rule: str
    config: Dict[str, Any]


@dataclass
class ModelResponse:
    """Model response with metadata"""
    content: str
    model: str
    cost: float
    latency: int


class ModelRouter:
    """
    Model Router
    Handles model selection based on routing configuration
    """

    def __init__(self, config: RoutingConfig):
        self.config = config
        self.request_count: Dict[str, int] = {}
        self.total_cost: float = 0.0

    def select_model(self, context: RequestContext) -> ModelSelection:
        """Select best model based on request context"""

        # Apply routing rules
        if self.config.routing_rules:
            for rule_name, rule in self.config.routing_rules.items():
                if self._matches_rule(context, rule):
                    return ModelSelection(
                        model=rule.models[0],
                        fallback=rule.fallback or rule.models[1:],
                        rule=rule_name,
                        config={
                            'max_tokens': rule.max_tokens,
                            'temperature': rule.temperature,
                            'streaming': rule.streaming
                        }
                    )

        # Default to primary/fallback
        return ModelSelection(
            model=self.config.primary or 'anthropic/claude-4.5-sonnet',
            fallback=self.config.fallback or [],
            rule='default',
            config={}
        )

    def _matches_rule(self, context: RequestContext, rule: RoutingRule) -> bool:
        """Check if request matches routing rule"""
        if not rule.classifier:
            return False

        conditions = rule.classifier.get('conditions', [])

        for condition in conditions:
            # Simple condition evaluation (extend as needed)
            if 'token_count <' in condition and context.token_count:
                threshold = int(condition.split('<')[1].strip())
                if context.token_count >= threshold:
                    return False

            if 'token_count >=' in condition and context.token_count:
                threshold = int(condition.split('>=')[1].split()[0])
                if context.token_count < threshold:
                    return False

            if 'task_type in' in condition and context.task_type:
                # Extract task types from condition
                allowed_types = condition.split('[')[1].split(']')[0]
                allowed_types = [t.strip().strip("'\"") for t in allowed_types.split(',')]
                if context.task_type not in allowed_types:
                    return False

        return True

    async def execute_with_fallback(
        self,
        context: RequestContext,
        api_client: 'OpenRouterClient'
    ) -> ModelResponse:
        """Execute request with fallback chain"""
        selection = self.select_model(context)
        models = [selection.model] + selection.fallback

        last_error = None

        for i, model in enumerate(models):
            try:
                print(f"Attempting model {i + 1}/{len(models)}: {model}")

                response = await self._execute_with_retry(
                    model,
                    context,
                    api_client,
                    selection.config
                )

                # Track metrics
                self._track_request(model, response.cost)

                return response

            except Exception as error:
                last_error = error
                print(f"Model {model} failed: {error}")

                if i == len(models) - 1:
                    # Last model in chain
                    raise Exception(f"All models failed. Last error: {last_error}")

                # Continue to next model
                continue

        raise Exception("Fallback chain exhausted")

    async def _execute_with_retry(
        self,
        model: str,
        context: RequestContext,
        api_client: 'OpenRouterClient',
        rule_config: Dict[str, Any]
    ) -> ModelResponse:
        """Execute request with retry logic"""
        max_attempts = self.config.retry.max_attempts
        delay = self.config.retry.delay_ms / 1000  # Convert to seconds

        for attempt in range(1, max_attempts + 1):
            try:
                response = await api_client.chat(
                    model=model,
                    messages=context.messages,
                    max_tokens=rule_config.get('max_tokens', 2000),
                    temperature=rule_config.get('temperature', 0.7),
                    stream=rule_config.get('streaming', False)
                )

                return response

            except Exception as error:
                if attempt == max_attempts:
                    raise error

                print(f"Retry attempt {attempt}/{max_attempts} after {delay}s")
                time.sleep(delay)

                if self.config.retry.exponential_backoff:
                    delay *= 2

        raise Exception("Max retries exceeded")

    def _track_request(self, model: str, cost: float):
        """Track request metrics"""
        self.request_count[model] = self.request_count.get(model, 0) + 1
        self.total_cost += cost

        # Check budget alerts
        if self.config.cost_tracking and self.config.cost_tracking.alert_threshold_usd:
            if self.total_cost >= self.config.cost_tracking.alert_threshold_usd:
                print(f"⚠️ Cost alert: ${self.total_cost:.2f} >= ${self.config.cost_tracking.alert_threshold_usd}")

    def get_stats(self) -> Dict[str, Any]:
        """Get routing statistics"""
        total_requests = sum(self.request_count.values())

        return {
            'total_requests': total_requests,
            'total_cost': self.total_cost,
            'model_distribution': self.request_count,
            'avg_cost_per_request': self.total_cost / total_requests if total_requests > 0 else 0
        }


class OpenRouterClient:
    """Mock OpenRouter API client (replace with actual implementation)"""

    async def chat(
        self,
        model: str,
        messages: List[Dict[str, str]],
        max_tokens: int,
        temperature: float,
        stream: bool
    ) -> ModelResponse:
        """Execute chat completion request"""
        # TODO: Implement actual API call
        return ModelResponse(
            content=f"Response from {model}",
            model=model,
            cost=0.001,
            latency=500
        )


def load_config_from_file(filepath: str) -> RoutingConfig:
    """Load routing configuration from JSON file"""
    with open(filepath, 'r') as f:
        data = json.load(f)

    # Convert to RoutingConfig
    retry_config = RetryConfig(**data.get('retry', {}))

    routing_rules = None
    if 'routing_rules' in data:
        routing_rules = {
            name: RoutingRule(**rule)
            for name, rule in data['routing_rules'].items()
        }

    cost_tracking = None
    if 'cost_tracking' in data:
        cost_tracking = CostTrackingConfig(**data['cost_tracking'])

    monitoring = None
    if 'monitoring' in data:
        monitoring = MonitoringConfig(**data['monitoring'])

    return RoutingConfig(
        strategy=RoutingStrategy(data['strategy']),
        description=data['description'],
        version=data.get('version', '1.0.0'),
        primary=data.get('primary'),
        fallback=data.get('fallback'),
        timeout=data.get('timeout', 5000),
        retry=retry_config,
        on_error=data.get('on_error', 'fallback'),
        routing_rules=routing_rules,
        cost_tracking=cost_tracking,
        monitoring=monitoring
    )


# Example usage
async def example_usage():
    """Example usage of ModelRouter"""

    # Load configuration
    config = load_config_from_file('balanced-routing.json')

    # Create router
    router = ModelRouter(config)

    # Create API client
    api_client = OpenRouterClient()

    # Execute request
    context = RequestContext(
        messages=[{'role': 'user', 'content': 'Hello!'}],
        token_count=100,
        task_type='classification'
    )

    try:
        response = await router.execute_with_fallback(context, api_client)
        print('Response:', response)
        print('Stats:', router.get_stats())
    except Exception as error:
        print('Error:', error)


if __name__ == '__main__':
    import asyncio
    asyncio.run(example_usage())
